src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -712,23 +711,33 @@ ArrayJoinActionPtr SelectQueryExpressionAnalyzer::appendArrayJoin(ExpressionActi
     return array_join;
 }
 
-bool SelectQueryExpressionAnalyzer::appendJoinLeftKeys(ExpressionActionsChain & chain, bool only_types)
+bool SelectQueryExpressionAnalyzer::appendJoinLeftKeys(ExpressionActionsChain & chain, bool only_types, Block & block)
 {
     ExpressionActionsChain::Step & step = chain.lastStep(columns_after_array_join);
 
     getRootActions(analyzedJoin().leftKeysList(), only_types, step.actions());
+    ExpressionActionsPtr actions = std::make_shared<ExpressionActions>(step.actions());
+    actions->execute(block);
     return true;
 }
 
-JoinPtr SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain)
+JoinPtr
+SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain, const Block & sample_block, ActionsDAGPtr & before_join_dag)
 {
-    JoinPtr table_join = makeTableJoin(*syntax->ast_join);
+    JoinCommon::JoinConvertActions converting_actions;
+    JoinPtr table_join = makeTableJoin(*syntax->ast_join, sample_block, converting_actions);
+    if (converting_actions.needConvert())
+    {
+        syntax->analyzed_join->setConvertedRightType(converting_actions.right_target_types);
 
-    ExpressionActionsChain::Step & step = chain.lastStep(columns_after_array_join);
+        before_join_dag = ActionsDAG::merge(std::move(*before_join_dag->clone()), std::move(*converting_actions.left_actions->clone()));
 
-    chain.steps.push_back(std::make_unique<ExpressionActionsChain::JoinStep>(
-            syntax->analyzed_join, table_join, step.getResultColumns()));
+        chain.steps.push_back(std::make_unique<ExpressionActionsChain::ExpressionActionsStep>(converting_actions.left_actions));
```

Comment:
Here we change actions of previous step **and** add a new step to chain (which is not used anyway).
Changing of `before_join_dag` is not good here, because those actions are already in chain. It means, that `getResultColumns` for the last step will be changed - which may be unusual.

The correct thing to do is:
* add `converting_actions.left_actions` to chain (as it is already done. before_join_dag shouldn't be changed)
* return pair `(converting_actions.left_actions, JoinPtr)` from `appendJoin`
* add separate step to `ExpressionAnalysisResult` (something like `converting_join_types` between `before_join` and `join`
* add another one `ExpressionStep` before adding join (probably here: https://github.com/ClickHouse/ClickHouse/blob/a2ceca3fa18f1df125de02c7c09e3999f5df8fce/src/Interpreters/InterpreterSelectQuery.cpp#L978-L979)

Reply:
Yes, it's really simplify code. Now actions for converting stored in [converting_join_columns](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-5452971ae18972aa669265008a29b3732a99d9415f5b3af10d573c0807667a22R981)